Port tests for changes to values to expect tests

Siddhi-agg opened the pull request:
Aims to close #33. This is my initial work on the issue. I have opened this PR to record my progress and display it for ease in reviewing.

Siddhi-agg commented:
@NathanReb I had a question. When I execute ```dune runtest``` I get the following error : 
```
......................................{
                  desc = Tconstr (Path.Pident "int", [], ref []);
                  level = 0;
                  scope =0;
                  id = 1;
                };
Error: Cannot create values of the private type Types.transient_expr
```
Then, when I change the code to:
```
 let a = 0 in
  let signature1=
  [
   .... 
        type_manifest = Some {
                  desc = Tconstr (Path.Pident "int", [], ref []);
                  level = 0;
                  scope = Some a;
                  id = 1;
                };
``` 
It still gives the same error. Shouldn't changing the scope from 0 to Some a (0) change its type from ```int``` to ```int option``` and thus change the whole expression's type to type_expr?

harshey1103 commented:
Hi @Siddhi-agg, I encountered a similar error to the one you mentioned. Since I am responsible for implementing the type part. You can refer [here](https://github.com/NathanReb/ocaml-api-watch/pull/36/files#diff-e629cb7eb02891f07b57494e4d7ad296e9545ce6b763e4178cae4165e134bd21R18) for the fix.

NathanReb commented:
Indeed @Siddhi-agg, your error was that you were trying to build a value of a private type.

They are documented [here](https://v2.ocaml.org/manual/privatetypes.html) if you want to take a look but in short, private types are kind of "read-only" types as in you can pattern match over their structure as if they were public but you cannot directly create a value of that type, you have to use API functions to do so just as if they were abstract.

E.g. if you have the following:
```ocaml
type t = private
  | A
  | B of int
  
val a : unit -> t
val b : int -> t
```

That means you can write the following code:
```ocaml
match t with
| A -> ...
| B i -> ...
```

but you cannot write:
```ocaml
let x = B 10
```

You will have to instead use the function provided above:
```ocaml
let x = b 10
```

Siddhi-agg commented:
> They are documented [here](https://v2.ocaml.org/manual/privatetypes.html) if you want to take a look but in short, private types are kind of "read-only" types as in you can pattern match over their structure as if they were public but you cannot directly create a value of that type, you have to use API functions to do so just as if they were abstract.

Yeah, I read that and used ```create``` function from to get a declare a ```type_expr```. @harshey1103 's implementation helped a lot in that. Thanks :)

I am currently working on why the ```diff_interface``` isn't detecting the modified values. I also noticed that @harshey1103 and @MungaSoftwiz are experiencing the same issue in their respective tests. Would be glad to help them if I solve the issue.

MungaSoftwiz commented:
> > They are documented [here](https://v2.ocaml.org/manual/privatetypes.html) if you want to take a look but in short, private types are kind of "read-only" types as in you can pattern match over their structure as if they were public but you cannot directly create a value of that type, you have to use API functions to do so just as if they were abstract.
> 
> Yeah, I read that and used `create` function from to get a declare a `type_expr`. @harshey1103 's implementation helped a lot in that. Thanks :)
> 
> I am currently working on why the `diff_interface` isn't detecting the modified values. I also noticed that @harshey1103 and @MungaSoftwiz are experiencing the same issue in their respective tests. Would be glad to help them if I solve the issue.

Yes that's an issue I have experienced too. From my PR I have used the expect tests as a placeholder for now. I've been trying to see how to navigate that issue better too. 
What's the output on your side please?

Siddhi-agg commented:
> What's the output on your side please?

@MungaSoftwiz On my side, I just get asked to change the expected output of the modify test to ```false```:
```File "tests/api-watch-diff/test.ml", line 1, characters 0-0:
diff --git a/_build/default/tests/api-watch-diff/test.ml b/_build/default/tests/api-watch-diff/test.ml.corrected
index 491825a..f3d351f 100644
--- a/_build/default/tests/api-watch-diff/test.ml
+++ b/_build/default/tests/api-watch-diff/test.ml.corrected
@@ -180,7 +180,7 @@ let%expect_test "modifying_a_value_test" =
       ~current:ref_signature
   in
   Format.printf "%b" result;
-  [%expect {|true|}]
+  [%expect {|false|}]
 
 (* Signature for remove_value.mli:
      > type t = int
```

MungaSoftwiz commented:
> > What's the output on your side please?
> 
> @MungaSoftwiz On my side, I just get asked to change the expected output of the modify test to `false`:
> 
> ```
> diff --git a/_build/default/tests/api-watch-diff/test.ml b/_build/default/tests/api-watch-diff/test.ml.corrected
> index 491825a..f3d351f 100644
> --- a/_build/default/tests/api-watch-diff/test.ml
> +++ b/_build/default/tests/api-watch-diff/test.ml.corrected
> @@ -180,7 +180,7 @@ let%expect_test "modifying_a_value_test" =
>        ~current:ref_signature
>    in
>    Format.printf "%b" result;
> -  [%expect {|true|}]
> +  [%expect {|false|}]
>  
>  (* Signature for remove_value.mli:
>       > type t = int
> ```

Interesting @Siddhi-agg! It's surprising that a change in types in the modules outputs a more verbose exception.
This bags the question, why is the change of a type in a module causing the particular backtrace compared to values?

shonfeder commented on tests/api-watch-diff/test.ml:
```diff
@@ -2,3 +2,196 @@ let%expect_test "test_diff_interface" =
+        type_manifest =
+          Some
+            (Transient_expr.type_expr
+               (Transient_expr.create (Tvar (Some "int")) ~level:0 ~scope:0
```
I think you are hitting the same problem I explained here https://github.com/NathanReb/ocaml-api-watch/pull/36#discussion_r1529470985 #37 

The gist is that these should be type constructors (`Tconst`) and not type variables (`Tvar`) :)

Siddhi-agg replied:
@shonfeder Thanks for this approach! I think the it worked. Also, we would have to use ```Ident.create_predef``` instead of ```Ident.create_persistent``` when giving the path to ```Tconstr``` because the ```Global``` identifier doesn't let the test detect the changes made to the ```value```.

NathanReb commented on tests/api-watch-diff/test.ml:
```diff
@@ -2,3 +2,214 @@ let%expect_test "test_diff_interface" =
+                    Transient_expr.type_expr
+                      (Transient_expr.create
+                         (Tconstr
+                            (Path.Pident (Ident.create_predef "t"), [], ref Mnil))
```
I think you shouldn't be using `create_predef` here. `t` is not a predefined OCaml type.

NathanReb commented on tests/api-watch-diff/test.ml:
```diff
@@ -2,3 +2,214 @@ let%expect_test "test_diff_interface" =
+            (Transient_expr.type_expr
+               (Transient_expr.create
+                  (Tconstr
+                     (Path.Pident (Ident.create_predef "string"), [], ref Mnil))
```
I think you can directly use `Predef.path_string` here. Same apply to `int` or any predefined ocaml type.

NathanReb commented on tests/api-watch-diff/test.ml:
```diff
@@ -2,3 +2,214 @@ let%expect_test "test_diff_interface" =
+        type_private = Public;
+        type_manifest =
+          Some
+            (Transient_expr.type_expr
```
I think you can use `Typing.create_expr` directly here instead of `Transient_expr. (type_exr (create ...))`.

NathanReb reviewed (commented):
Thanks! This is very useful work you're doing as it helps us dive into the compiler's representation of interfaces.

It also helped us figure out that in its current form it is neither a satisfying nor a scalable way to write tests and we need helpers to build signatures.

I still think it's worth going through with this before #35. I left a few comments for some improvements but besides that it looks good.

Siddhi-agg commented:
> It also helped us figure out that in its current form it is neither a satisfying nor a scalable way to write tests and we need helpers to build signatures.

@NathanReb I agree. When I was building signatures for tests, even I thought that was a tedious and verbose process to do manually for the various tests this project may have in the future. Having a helper library to automate generating signatures would be nice!

shonfeder commented on tests/api-watch-diff/test.ml:
```diff
@@ -2,3 +2,198 @@ let%expect_test "test_diff_interface" =
+      },
+      Exported )
+
+(* Signature for file ref.mli:
+      > type t = int
+      > type unused_type = string
+      > val f : t -> string *)
+
+let ref_signature = [ t_sig; unused_type_sig; val_f_sig ]
```
This is a really nice approach, and it saves having to write a bunch of duplicate code. It makes your tests very easy to read.

Siddhi-agg replied:
Thanks @shonfeder :)

shonfeder reviewed (approved):
Looks great!
